test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,2071 @@
+
+    /* cleanup buffers */
+    if (dtpi) {
+        free(dtpi);
+    }
+    if (buf) {
+        free(buf);
+    }
+
+    goto fn_exit;
+}
+
+int DTPI_Subarray_c_create(DTPI_Bound_e bound, struct DTPI_Par *par, DTP_t dtp)
+{
+    int err = DTP_SUCCESS;
+    int basic_type_size;
+    int len;
+    int obj_idx;
+    MPI_Aint lb, extent;
+    MPI_Datatype basic_type;
+    void *buf = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+    char basic_type_name[BASIC_TYPE_NAME_MAXLEN] = { 0 };
+    DTPI_t *dtpi = NULL;
+
+    /* get type signature for pool */
+    basic_type = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type;
+
+    /* get object index in the pool */
+    obj_idx = par->user.obj_idx;
+
+    err = MPI_Type_size(basic_type, &basic_type_size);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    /* allocate space for private datatype info */
+    DTPI_OBJ_ALLOC_OR_FAIL(dtpi, sizeof(*dtpi));
+
+    dtpi->u.subarray.arr_starts[0] = par->core.type_count / 8;  /* use fix offset along x */
+    dtpi->u.subarray.arr_starts[1] = par->core.type_stride - par->core.type_blklen;
+    dtpi->u.subarray.arr_sizes[0] = par->core.type_count + dtpi->u.subarray.arr_starts[0];
+    dtpi->u.subarray.arr_sizes[1] = par->core.type_stride;
+    dtpi->u.subarray.arr_subsizes[0] = par->core.type_count;
+    dtpi->u.subarray.arr_subsizes[1] = par->core.type_blklen;
+    dtpi->u.subarray.order = MPI_ORDER_C;
+
+    err = MPI_Type_create_subarray(2,
+                                   dtpi->u.subarray.arr_sizes,
+                                   dtpi->u.subarray.arr_subsizes,
+                                   dtpi->u.subarray.arr_starts,
+                                   dtpi->u.subarray.order,
+                                   basic_type, &dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    err = MPI_Type_commit(&dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    dtp->DTP_obj_array[obj_idx].DTP_obj_count = 1;
+
+    err = MPI_Type_get_extent(dtp->DTP_obj_array[obj_idx].DTP_obj_type, &lb, &extent);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    /* allocate buffer */
+    DTPI_OBJ_ALLOC_OR_FAIL(buf, lb + extent);
+
+    /* recompute stride and displacement info for buffer initialization */
+    par->core.type_displ =
+        ((dtpi->u.subarray.arr_sizes[1] * dtpi->u.subarray.arr_starts[0]) +
+         dtpi->u.subarray.arr_starts[1]) * basic_type_size;
+    par->core.type_extent = (extent - par->core.type_displ) / basic_type_size;
+
+    /* initialize buffer with requested DTPI_Par */
+    DTPI_Type_init_buf(par, basic_type, buf);
+
+    /* initialize private datatype info data */
+    dtpi->obj_type = DTPI_OBJ_TYPE__SUBARRAY_C;
+    dtpi->type_basic_size = basic_type_size;
+    dtpi->type_extent = extent - par->core.type_displ;
```

Comment:
Because the check buffer function (which is the same for every datatype) takes the pointer to the first valid element in the buffer (subarray has a non zero start offset for data i.e., `arr_starts[]`) and iterates for all the remaining elements (given by `extent / basic_type_size`). `par->core.type_displs` accounts for the start offset in the buffer thus if this is non zero the `extent` (which is normally the buffer size, in fact `lb` is almost everytime zero) should be resized to account for lesser elements.

Reply:
This is very confusing. `type_extent` should just mean the extent of the new datatype as defined in standard.

Reply:
I will fix it